nixos/modules/config/networking.nix

```diff
@@ -188,11 +214,23 @@ in
 
         # /etc/hosts: Hostname-to-IP mappings.
         "hosts".text =
+          let oneToString = set : ip : ip + " " + concatStringsSep " " ( getAttr ip set );
+              allToString = set : concatMapStringsSep "\n" ( oneToString set ) ( attrNames set );
+              userLocalHosts = optionalString
+                ( builtins.hasAttr "127.0.0.1" cfg.hosts )
+                ( concatStringsSep " " ( remove "localhost" cfg.hosts."127.0.0.1" ));
+              userLocalHosts6 = optionalString
+                ( builtins.hasAttr "::1" cfg.hosts )
+                ( concatStringsSep " " ( remove "localhost" cfg.hosts."::1" ));
+              otherHosts = allToString ( removeAttrs cfg.hosts [ "127.0.0.1" "::1" ]);
+              maybeFQDN = optionalString ( cfg.fqdn != null ) cfg.fqdn;
```

Comment:
@aszlig its ability to merge worries me. IIRC, in case of merging two different configurations for same IP `canonical` will be overwritten, while it should be demoted to alias (with possible warning to user). It seems to be rare case, but still legal and should be predictable. Beside that, I like your variant.